packages/@react-spectrum/tabs/src/Tabs.tsx

```diff
@@ -34,64 +36,140 @@ export function Tabs<T extends object>(props: SpectrumTabsProps<T>) {
+        let shouldCollapse = direction === 'rtl' ? farEdgeLastTab < farEdgeTabList : farEdgeTabList < farEdgeLastTab;
+
+        if (shouldCollapse) {
+          return true;
+        } else {
+          return false;
+        }
+      }
+    };
+
+    if (orientation !== 'vertical') {
+      setCollapse(function* () {
+        // Make Tabs render in non-collapsed state
+        yield false;
+
+        // Compute if Tabs should collapse and update
+        yield computeShouldCollapse();
+      });
+    }
+  }, [ref, wrapperRef, direction, orientation, setCollapse]);
+
+  useEffect(() => {
+    checkShouldCollapse();
+  }, [props.children, checkShouldCollapse]);
+
+  useResizeObserver({ref: wrapperRef.current && wrapperRef, onResize: checkShouldCollapse});
+
+  let tablist = (
+    <TabList
+      {...tabListProps}
+      ref={ref}
+      orientation={orientation}
+      density={density}
+      isQuiet={isQuiet}
+      isDisabled={isDisabled}
+      state={state}
+      selectedTab={selectedTab} />
+  );
+
+  if (collapse && orientation !== 'vertical') {
+    tablist = (
+      <TabsPicker
+        {...props}
+        state={state} />
+    );
+  }
+
+  if (orientation !== 'vertical') {
+    tablist = (
+      <div
+        ref={wrapperRef}
+        className={classNames(
+          styles,
+          'spectrum-Tabs--collapsible'
+        )}>
+        {tablist}
+      </div>
+    );
+  }
 
   return (
     <div
       {...styleProps}
       className={classNames(
-        tabsStyles,
-        'react-spectrum-TabPanel',
-        `react-spectrum-TabPanel--${orientation}`,
+        styles,
+        'spectrum-TabsPanel',
+        `spectrum-TabsPanel--${orientation}`,
         styleProps.className
       )}>
-      <div
-        {...styleProps}
-        {...tabListProps}
-        ref={ref}
-        className={classNames(
-          styles,
-          'spectrum-Tabs',
-          `spectrum-Tabs--${orientation}`,
-          {'spectrum-Tabs--quiet': isQuiet},
-          density ? `spectrum-Tabs--${density}` : '',
-          styleProps.className
-        )}>
-        {[...state.collection].map((item) => (
-          <Tab key={item.key} item={item} state={state} isDisabled={isDisabled} orientation={orientation} />
-        ))}
-        {selectedTab && <TabLine orientation={orientation} selectedTab={selectedTab} />}
-      </div>
-      <div {...tabPanelProps} className="react-spectrum-TabPanel-body">
+      {tablist}
+      <div {...tabPanelProps}>
         {state.selectedItem && state.selectedItem.props.children}
       </div>
     </div>
   );
 }
 
-interface TabProps<T> extends DOMProps, StyleProps {
+interface TabProps<T> extends DOMProps {
   item: Node<T>,
   state: SingleSelectListState<T>,
   isDisabled?: boolean,
   orientation?: Orientation
 }
 
 export function Tab<T>(props: TabProps<T>) {
-  let {item, state, isDisabled: propsDisabled, ...otherProps} = props;
-  let {styleProps} = useStyleProps(otherProps);
```

Comment:
removed the `styleProps`/`styleProps.className` from Tab/Tablist since it feels like those should only affect the Tabs wrapping div